build_node/utils/codenotary.py

```diff
@@ -0,0 +1,76 @@
+import time
+import typing
+
+from cas_wrapper import CasWrapper
+
+from build_node.models import Task, Artifact
+from build_node.utils.rpm_utils import get_rpm_metadata
+
+__all__ = [
+    'notarize_build_artifacts',
+]
+
+
+def notarize_build_artifacts(
+    task: Task,
+    build_artifacts: typing.List[Artifact],
+    cas_client: CasWrapper,
+    build_host: str,
+):
+    cas_metadata = {
+        'build_id': task.build_id,
+        'build_host': build_host,
+        'build_arch': task.arch,
+        'built_by': task.created_by.full_name,
+        'sbom_api': '0.1',
+    }
+    if task.is_alma_source() and task.alma_commit_cas_hash:
+        cas_metadata['alma_commit_sbom_hash'] = task.alma_commit_cas_hash
+    if task.ref.git_ref:
+        cas_metadata.update({
+            'source_type': 'git',
+            'git_url': task.ref.url,
+            'git_ref': task.ref.git_ref,
+            'git_commit': task.ref.git_commit_hash,
+        })
+    else:
+        srpm = next((
+            artifact for artifact in build_artifacts
+            if artifact.name.endswith('.src.rpm')
+        ), None)
+        if srpm:
+            hdr = get_rpm_metadata(srpm.path)
+            srpm_nevra = (
+                f"{hdr['epoch']}:{hdr['name']}-{hdr['version']}-"
+                f"{hdr['release']}.src"
+            )
+            srpm_sha256 = task.srpm_hash if task.srpm_hash else srpm.sha256
+            cas_metadata.update({
+                'source_type': 'srpm',
+                'srpm_url': task.ref.url,
+                'srpm_sha256': srpm_sha256,
+                'srpm_nevra': srpm_nevra,
+            })
+
+    notarized_artifacts = {}
+    artifact_paths = [artifact.path for artifact in build_artifacts]
+    all_artifacts_is_notarized, artifacts = cas_client.notarize_artifacts(
+        artifact_paths=artifact_paths,
+        metadata=cas_metadata,
+    )
+    notarized_artifacts.update(artifacts)
+    # ensure that all artifacts notarized
+    while not all_artifacts_is_notarized:
+        artifact_paths = [
+            artifact.path
+            for artifact in build_artifacts
+            if artifact.path not in notarized_artifacts
+        ]
+        all_artifacts_is_notarized, artifacts = cas_client.notarize_artifacts(
+            artifact_paths=artifact_paths,
+            metadata=cas_metadata,
+        )
+        notarized_artifacts.update(artifacts)
+        time.sleep(5)
```

Comment:
Unless I've lost something, @AvoidMe's question is still relevant here, right?

`It feels like it can be infinity loop some day`

If so, adding a sleep helps to give some time until we check the next time, but what if, in the end, some artifact can't be notarized? Could be worth adding a timeout/max retries approach here?

Reply:
And now that I'm seeing this more in context, is this subject to use an async/await approach?
As I said, maybe I'm missing some context here, so nevermind if my review comment doesn't make sense.

Reply:
I removed this loop, locally it's works fine, I'm think that if we would have problems with some non notarized artifacts, we can make renotarization for failed artifacts
About async/await approach, `notarize_artifacts` func use `ThreadPoolExecutor` (https://git.almalinux.org/almalinux/cas_wrapper/src/tag/0.0.2/cas_wrapper.py#L156), so there is no need to use asyncio, also our build node works in synchronous mode

Reply:
Right, we're using a threaded approach, nevermind :)
And yeah, I totally agree with you, let's see if we run into such situation in the future and if so, we'll think a way to ensure that everything is notarized. Thanks!!

Reply:
I guess we can resolve this unless @AvoidMe thinks otherwise.